File: src/_004/kotlin/Solution.kt
package _004.kotlin

/**
 * @author relish
 * @since 2018/10/20
 */
class Solution {
    fun findMedianSortedArrays(nums1: IntArray, nums2: IntArray): Double {
        val len = nums1.size + nums2.size
        return if (len % 2 == 0) {
            (fix(nums1, 0, nums2, 0, len / 2) + fix(nums1, 0, nums2, 0, len / 2 + 1)) / 2.0
        } else {
            fix(nums1, 0, nums2, 0, len / 2 + 1).toDouble()
        }
    }

    fun fix(nums1: IntArray, n: Int, nums2: IntArray, m: Int, k: Int): Int {
        if (n >= nums1.size) return nums2[m + k - 1]
        if (m >= nums2.size) return nums1[n + k - 1]
        if (k == 1) return Math.min(nums1[n], nums2[m])

        val i1 = n + k / 2 - 1
        val i2 = m + k / 2 - 1
        val v1 = if (i1 < nums1.size) nums1[i1] else Int.MAX_VALUE
        val v2 = if (i2 < nums2.size) nums2[i2] else Int.MAX_VALUE
        return if (v1 < v2) {
            fix(nums1, n + k / 2, nums2, m, k - k / 2)
        } else {
            fix(nums1, n, nums2, m + k / 2, k - k / 2)
        }
    }
}

fun main(args: Array<String>) {
    val solution = Solution()
    println(solution.findMedianSortedArrays(intArrayOf(2), intArrayOf(0)))
    println(solution.findMedianSortedArrays(intArrayOf(1, 2), intArrayOf(3, 4)))
    println(solution.findMedianSortedArrays(intArrayOf(1, 2), intArrayOf(-1, 3)))
}

File: src/_004/Solution.js
var findMedianSortedArrays = function(nums1, nums2) {
    var l1 = nums1.length;
    var l2 = nums2.length;
    var i = 0, j = 0, k = 0;
    var midIndex = Math.ceil((l1+l2) / 2)
    var midNum1,midNum2,midNum
    if((l1+l2)%2 === 0) {
       while(k <= midIndex) {
            k++;
			if(i > l1 - 1) {
				midNum1  = nums2[midIndex-k +j]
				midNum2  =nums2[midIndex-k +j + 1]
				break;
			}
			if(j > l2 - 1) {
				midNum1  = nums2[midIndex-k +i]
				midNum2  =nums2[midIndex-k +i + 1]
				break;
			}
            if(nums1[i]>nums2[j]) {
                if(k === midIndex) {
                    midNum1 = nums2[j];
                } else if(k === (midIndex+1)){
					midNum2 = nums2[j];
				}
                j++;
            } else {
                if(k === midIndex) {
                    midNum1 = nums1[i];
                } else if(k === (midIndex+1)){
					midNum2 = nums1[i];
				}
                i++;
            }
        }
	midNum = (midNum1 + midNum2) / 2
    } else {
        while(k < midIndex) {
            k++;
            if(nums1[i]>nums2[j]) {
                j++;
                if(k === midIndex) {
                    midNum = nums2[j-1];
                }
            } else {
                i++;
                if(k === midIndex) {
                    midNum = nums1[i-1];
                }
            }
        }
    }
    return midNum
};
findMedianSortedArrays([1,2,4,6],[3,4])